docs/source/reference/public.rst

```diff
@@ -0,0 +1,6 @@
+.. module:: optuna
+
+Public API
```

Comment:
Hm, cannot come up with a definite idea and just naming a few (without really motivating them) but how about "Core", "Basic", "Common"? The fact that it's "public" is already implied and could be slightly confusing as one might think that the other parts are non-public, which is not the case. This is a minor but "API" is a bit redundant as it's mentioned in the section header "API reference". 🤔

Reply:
Using module names might more straight forward but maybe beyond the scope of this PR.

Reply:
> Using module names might more straight forward but maybe beyond the scope of this PR.
 
I agree with this idea, but it will change all structure of reference documentation. I will hold a discussion about it later.